File: CtCI/Bubble-Sort.py
n = int(raw_input().strip())
a = map(int, raw_input().strip().split(' '))

def bubbleSort(array):
    numSwaps,totSwaps = 0,0
    
    while True:
        for idx in range(len(array)-1):
            if array[idx]>array[idx+1]:
                x = array[idx]
                array[idx]=array[idx+1]
                array[idx+1]=x
                numSwaps+=1
                totSwaps+=1
        if numSwaps == 0:
            break
        else:
            numSwaps = 0
    
    print "Array is sorted in %i swaps." % totSwaps
    print "First Element: %i" % array[0]
    print "Last Element: %i" % array[-1]

bubbleSort(a)

File: CtCI/Bubble-Sort.c
#include <math.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <assert.h>
#include <limits.h>
#include <stdbool.h>

// Did this again for practice with C.

int bubbleSort(int *a, int n);

int bubbleSort(int *a, int n) {
    int totalSwaps = 0;
    int numSwaps;
    do {
        numSwaps = 0;
        for (int i=0; i<n-1; i++) {
            if (a[i]>a[i+1]) {
                int temp = a[i];
                a[i] = a[i+1];
                a[i+1] = temp;
                totalSwaps++;
                numSwaps++;
            }
        }
    } while(numSwaps != 0);
    return totalSwaps;
}

int main(){
    int n; 
    scanf("%d",&n);
    int *a = malloc(sizeof(int) * n);
    for(int a_i = 0; a_i < n; a_i++){
       scanf("%d",&a[a_i]);
    }
    int totSwaps = bubbleSort(a, n);
    printf("Array is sorted in %d swaps.\n", totSwaps);
    printf("First Element: %d\n", a[0]);
    printf("Last Element: %d\n",a[n-1]);
    return 0;
}
